pkg/innerring/config.go

```diff
@@ -33,258 +24,189 @@ const (
 	}
-
-	c.BlockInterval, err = parseConfigDurationPositive(v, cfgPathFSChainLocalConsensus+".time_per_block", "block interval")
-	if err != nil && !errors.Is(err, errMissingConfig) {
-		return c, err
+	if cfgFSChainLocalConsensus.Committee.Len() == 0 {
+		return fmt.Errorf("empty committee members '%s'", committeeKey)
 	}
 
-	traceableChainLength, err := parseConfigUint64Range(v, cfgPathFSChainLocalConsensus+".max_traceable_blocks", "traceable chain length", 1, math.MaxUint32)
-	if err != nil && !errors.Is(err, errMissingConfig) {
-		return c, err
+	err = checkDurationPositive(cfgFSChainLocalConsensus.TimePerBlock, cfgPathFSChainLocalConsensus+".time_per_block")
+	if err != nil {
+		return err
 	}
-	c.TraceableChainLength = uint32(traceableChainLength)
 
-	c.SeedNodes, err = parseConfigAddressesTCP(v, cfgPathFSChainLocalConsensus+".seed_nodes", "seed nodes", p2pDefaultListenPort)
-	if err != nil && !errors.Is(err, errMissingConfig) {
-		return c, err
+	cfgFSChainLocalConsensus.SeedNodes, err = parseConfigAddressesTCP(cfgFSChainLocalConsensus.SeedNodes,
+		cfgPathFSChainLocalConsensus+".seed_nodes", p2pDefaultListenPort)
+	if err != nil {
+		return err
 	}
 
 	var hardForksKey = cfgPathFSChainLocalConsensus + ".hardforks"
-	if v.IsSet(hardForksKey) {
-		c.HardForks, err = parseConfigMapUint32(v, hardForksKey, "hard forks", math.MaxUint32)
-		if err != nil {
-			return c, err
+	if cfg.IsSet(hardForksKey) {
+		for k := range cfgFSChainLocalConsensus.Hardforks.Name {
+			if k == "" {
+				return fmt.Errorf("missing hardforks name '%s'", hardForksKey)
+			}
 		}
 	}
 
-	p2pNotaryRequestPayloadPoolSize, err := parseConfigUint64Range(v, cfgPathFSChainLocalConsensus+".p2p_notary_request_payload_pool_size",
-		"Size of the node's P2P Notary request payloads memory pool", 1, math.MaxUint32)
-	if err != nil && !errors.Is(err, errMissingConfig) {
-		return c, err
-	}
-	c.P2PNotaryRequestPayloadPoolSize = uint32(p2pNotaryRequestPayloadPoolSize)
+	if cfg.IsSet(cfgPathFSChainLocalConsensus + ".validators_history") {
+		committeeSize := uint32(cfgFSChainLocalConsensus.Committee.Len())
 
-	var validatorsHistoryKey = cfgPathFSChainLocalConsensus + ".validators_history"
-	if v.IsSet(validatorsHistoryKey) {
-		c.ValidatorsHistory = make(map[uint32]uint32)
-		committeeSize := uint64(c.Committee.Len())
-		err = parseConfigMap(v, validatorsHistoryKey, "validators history", func(name string, val any) error {
-			height, err := strconv.ParseUint(name, 10, 32)
-			if err != nil {
-				return fmt.Errorf("parse unsigned integer: %w", err)
+		for k, v := range cfgFSChainLocalConsensus.ValidatorsHistory.Height {
+			if k%committeeSize != 0 {
+				return fmt.Errorf("height %d is not a multiple of the %q size", k, committeeKey)
 			}
-
-			if height%committeeSize != 0 {
-				return fmt.Errorf("height %d is not a multiple of the %q size", height, committeeKey)
+			if v <= 0 {
+				return fmt.Errorf("value %d is out of allowable range", v)
```

Comment:
"is not a positive number" and k is even more helpful here for the user (as well as pointing to `ValidatorsHistory`).